addon/resolver.js

```diff
@@ -0,0 +1,370 @@
+  },
+
+  normalize: function(fullName) {
+    return this._normalizeCache[fullName] || (this._normalizeCache[fullName] = this._normalize(fullName));
+  },
+
+  _normalize: function(fullName) {
+    // replace `.` with `/` in order to make nested controllers work in the following cases
+    // 1. `needs: ['posts/post']`
+    // 2. `{{render "posts/post"}}`
+    // 3. `this.render('posts/post')` from Route
+    var split = fullName.split(':');
+    if (split.length > 1) {
+      return split[0] + ':' + Ember.String.dasherize(split[1].replace(/\./g, '/'));
+    } else {
+      return fullName;
+    }
+  },
+
+  pluralize: function(type) {
+    return this.pluralizedTypes[type] || (this.pluralizedTypes[type] = type + 's');
+  },
+
+  podBasedLookupWithPrefix: function(podPrefix, parsedName) {
+    var fullNameWithoutType = parsedName.fullNameWithoutType;
+
+    if (parsedName.type === 'template') {
+      fullNameWithoutType = fullNameWithoutType.replace(/^components\//, '');
+    }
+
+    return podPrefix + '/' + fullNameWithoutType + '/' + parsedName.type;
+  },
+
+  podBasedModuleName: function(parsedName) {
+    var podPrefix = this.namespace.podModulePrefix || this.namespace.modulePrefix;
+
+    return this.podBasedLookupWithPrefix(podPrefix, parsedName);
+  },
+
+  podBasedComponentsInSubdir: function(parsedName) {
+    var podPrefix = this.namespace.podModulePrefix || this.namespace.modulePrefix;
+    podPrefix = podPrefix + '/components';
+
+    if (parsedName.type === 'component' || parsedName.fullNameWithoutType.match(/^components/)) {
+      return this.podBasedLookupWithPrefix(podPrefix, parsedName);
+    }
+  },
+
+  mainModuleName: function(parsedName) {
+    // if router:main or adapter:main look for a module with just the type first
+    var tmpModuleName = parsedName.prefix + '/' + parsedName.type;
+
+    if (parsedName.fullNameWithoutType === 'main') {
+      return tmpModuleName;
+    }
+  },
+
+  defaultModuleName: function(parsedName) {
+    return parsedName.prefix + '/' +  this.pluralize(parsedName.type) + '/' + parsedName.fullNameWithoutType;
+  },
+
+  prefix: function(parsedName) {
+    var tmpPrefix = this.namespace.modulePrefix;
+
+    if (this.namespace[parsedName.type + 'Prefix']) {
+      tmpPrefix = this.namespace[parsedName.type + 'Prefix'];
+    }
+
+    return tmpPrefix;
+  },
+
+  /**
+
+   A listing of functions to test for moduleName's based on the provided
+   `parsedName`. This allows easy customization of additional module based
+   lookup patterns.
+
+   @property moduleNameLookupPatterns
+   @returns {Ember.Array}
+   */
+  moduleNameLookupPatterns: Ember.computed(function(){
+    return Ember.A([
+      this.podBasedModuleName,
+      this.podBasedComponentsInSubdir,
+      this.mainModuleName,
+      this.defaultModuleName
+    ]);
+  }),
+
+  findModuleName: function(parsedName, loggingDisabled){
+    var self = this;
+    var moduleName;
+
+    this.get('moduleNameLookupPatterns').find(function(item) {
```

Comment:
isn't needed for this PR. But I would prefer if we didn't rely on prototype extensions here. Instead just used a for-loop, as this little loop does get called quite often during resolution